examples/ExactlyOnce/Program.cs

```diff
@@ -14,789 +14,677 @@
 
         /// <summary>
-        ///     Name of the consumer group used in the reduce phase when
-        ///     consuming from the Topic_Words topic.
+        ///     Retry callback that sleeps for a configured period.
         /// </summary>
-        const string ConsumerGroup_Aggregate = "aggregate-consumer-group";
+        private void RetryAfterSleeping(string operation = null, KafkaException e = null)
+        {
+            Thread.Sleep(RetryInterval);
+        }
 
         /// <summary>
-        ///     Name of the consumer group used when loading count state
-        ///     from Topic_Counts.
+        ///     Retry callback that stops retries if
+        ///     the error is fatal or abortable or not retriable.
         /// </summary>
-        const string ConsumerGroup_LoadState = "load-state-consumer-group";
-
-
-        class FasterState : IDisposable
+        private void RetryIfNeeded(string operation, KafkaException e)
         {
-            public FasterState(Partition partition)
+            var TxnRequiresAbort = e is KafkaTxnRequiresAbortException;
+            var IsRetriable = e is KafkaRetriableException;
+            var IsFatal = e.Error.IsFatal;
+            logger.LogError($"{operation} Kafka Exception caught: '{e.Message}', IsFatal: {IsFatal}, TxnRequiresAbort: {TxnRequiresAbort}, IsRetriable: {IsRetriable}");
+            if (IsFatal || TxnRequiresAbort || !IsRetriable)
             {
-                this.Path = System.IO.Path.Join(System.IO.Path.GetTempPath(), "partition_" + (int)partition);
-                this.ObjectLog = Devices.CreateLogDevice(this.Path + "_obj.log");
-                this.Log = Devices.CreateLogDevice(this.Path + ".log");
-                this.Store = new FasterKV<string, int>(
-                    size: 1L << 20, // 1M cache lines of 64 bytes each = 64MB hash table
-                    logSettings: new LogSettings { LogDevice = this.Log, ObjectLogDevice = this.ObjectLog }
-                );
-                var funcs = new SimpleFunctions<string, int>((a, b) => a + b); // function used for read-modify-write (RMW).
-                this.Session = this.Store.NewSession(funcs);
+                throw e;
             }
+            RetryAfterSleeping();
+        }
 
-            string Path { get; set; }
-            IDevice Log { get; set; }
-            IDevice ObjectLog { get; set; }
-            public FasterKV<string, int> Store { get; set; }
-            public ClientSession<string, int, int, int, Empty, IFunctions<string, int, int, int, Empty>> Session { get; set; }
-
-            public void Dispose()
+        /// <summary>
+        ///     Retry callback that stops retries only if
+        ///     the error is fatal or abortable.
+        /// </summary>
+        private void RetryUnlessFatalOrAbortable(string operation, KafkaException e)
+        {
+            var TxnRequiresAbort = e is KafkaTxnRequiresAbortException;
```

Comment:
these lines are duplicated a bit.
I also don't think this function is needed, RetryIfNeeded will do.